File: nordgen-go/internal/generator/generator.go
package generator

import (
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"sort"
	"strconv"
	"strings"
	"sync"
	"sync/atomic"
	"time"

	"nordgen/internal/client"
	"nordgen/internal/models"
	"nordgen/internal/ui"
)

const fileNameMaxLength = 15

type FileWrite struct {
	Name    string
	Content []byte
}

type DirJob struct {
	Path  string
	Files []FileWrite
}

type Generator struct {
	client          *client.NordClient
	consoleManager  *ui.ConsoleManager
	Stats           models.GenerationStats
	outputDirectory string
}

func NewGenerator(c *client.NordClient, u *ui.ConsoleManager) *Generator {
	return &Generator{
		client:         c,
		consoleManager: u,
	}
}

func sanitizePathSegment(segment string) string {
	segment = strings.TrimSpace(segment)
	if segment == "" {
		return "unknown"
	}
	segment = strings.ToLower(segment)
	b := make([]byte, 0, len(segment))

	for i := 0; i < len(segment); i++ {
		c := segment[i]
		switch c {
		case '<', '>', ':', '"', '/', '\\', '|', '?', '*', '\x00', ' ':
			b = append(b, '_')
		default:
			b = append(b, c)
		}
	}
	res := string(b)
	for strings.HasSuffix(res, ".") {
		res = res[:len(res)-1]
	}
	if res == "" {
		return "unknown"
	}
	return res
}

func (g *Generator) Process(privateKey string, prefs models.UserPreferences) (string, error) {
	g.outputDirectory = fmt.Sprintf("nordvpn_configs_%s", time.Now().Format("20060102_150405"))

	g.consoleManager.StartStatus("Fetching data...")

	var wgData sync.WaitGroup
	var obsLat, obsLon float64
	var rawServers []models.RawServer
	var errServers error

	wgData.Add(2)
	go func() {
		defer wgData.Done()
		obsLat, obsLon, _ = g.client.GetGeo()
	}()
	go func() {
		defer wgData.Done()
		rawServers, errServers = g.client.GetServers()
	}()
	wgData.Wait()

	g.consoleManager.StopStatus()

	if errServers != nil || len(rawServers) == 0 {
		g.consoleManager.Fail("Failed to fetch server data")
		return "", fmt.Errorf("server data fetch failed")
	}
	g.consoleManager.Success("Fetched server data")

	g.consoleManager.StartStatus("Processing dataset...")

	allParsed := parseServers(rawServers, obsLat, obsLon, prefs.Groups, prefs.ExcludeDedicated)

	uniqueMap := make(map[string]models.Server, len(allParsed))
	for _, s := range allParsed {
		uniqueMap[s.Name] = s
	}

	uniqueServers := make([]models.Server, 0, len(uniqueMap))
	for _, s := range uniqueMap {
		uniqueServers = append(uniqueServers, s)
	}

	if len(uniqueServers) == 0 {
		g.consoleManager.StopStatus()
		g.consoleManager.Fail("No servers found matching the specified filters")
		return "", fmt.Errorf("no servers matched filters")
	}

	sort.Slice(uniqueServers, func(i, j int) bool {
		if uniqueServers[i].Load == uniqueServers[j].Load {
			return uniqueServers[i].Distance < uniqueServers[j].Distance
		}
		return uniqueServers[i].Load < uniqueServers[j].Load
	})

	g.Stats.Total = len(uniqueServers)

	type bestKey struct {
		combo   string
		country string
		city    string
	}
	bestMap := make(map[bestKey]models.Server, len(uniqueServers))
	for _, s := range uniqueServers {
		k := bestKey{s.Combo, s.Country, s.City}
		if _, exists := bestMap[k]; !exists {
			bestMap[k] = s
		}
	}
	g.Stats.Best = len(bestMap)

	bestServers := make([]models.Server, 0, len(bestMap))
	for _, s := range bestMap {
		bestServers = append(bestServers, s)
	}

	dirMap := make(map[string]*DirJob)
	g.buildJobs(uniqueServers, "configs", privateKey, prefs, dirMap)
	g.buildJobs(bestServers, "best_configs", privateKey, prefs, dirMap)

	g.consoleManager.StopStatus()
	g.consoleManager.Success("Dataset processed")

	var dirJobs []*DirJob
	totalFiles := 0
	for _, dj := range dirMap {
		dirJobs = append(dirJobs, dj)
		totalFiles += len(dj.Files)
	}

	errWrite := g.writeJobsParallel(dirJobs, totalFiles)

	if errWrite != nil {
		g.consoleManager.Fail(fmt.Sprintf("Failed to write configuration files: %v", errWrite))
		return "", fmt.Errorf("file write failed: %w", errWrite)
	}

	g.consoleManager.Success("Configuration files written")

	return g.outputDirectory, nil
}

func (g *Generator) buildJobs(servers []models.Server, subDir, privKey string, prefs models.UserPreferences, dirMap map[string]*DirJob) {
	interfaceBlock := []byte("[Interface]\nPrivateKey = " + privKey + "\nAddress = 10.5.0.2/16\nDNS = " + prefs.DNS + "\n\n[Peer]\n")
	keepaliveBlock := []byte("\nPersistentKeepalive = " + strconv.Itoa(prefs.Keepalive))
	pubKeyPrefix := []byte("PublicKey = ")
	allowedIPPrefix := []byte("\nAllowedIPs = 0.0.0.0/0, ::/0\nEndpoint = ")
	endpointSuffix := []byte(":51820")

	counts := make(map[string]int, len(servers))

	for _, s := range servers {
		countrySeg := sanitizePathSegment(s.Country)
		citySeg := sanitizePathSegment(s.City)

		fnameRoot := sanitizePathSegment(s.Name)
		if len(fnameRoot) > fileNameMaxLength {
			fnameRoot = fnameRoot[:fileNameMaxLength]
		}

		dir := filepath.Join(g.outputDirectory, subDir, s.Combo, countrySeg, citySeg)

		dj, exists := dirMap[dir]
		if !exists {
			dj = &DirJob{Path: dir}
			dirMap[dir] = dj
		}

		baseKey := filepath.Join(dir, fnameRoot)
		count := counts[baseKey]
		counts[baseKey] = count + 1

		fname := fnameRoot + ".conf"
		if count > 0 {
			fname = fnameRoot + "_" + strconv.Itoa(count) + ".conf"
		}

		endpoint := s.Hostname
		if prefs.UseIP {
			endpoint = s.Station
		}

		size := len(interfaceBlock) + len(pubKeyPrefix) + len(s.PublicKey) + len(allowedIPPrefix) + len(endpoint) + len(endpointSuffix) + len(keepaliveBlock)
		content := make([]byte, 0, size)
		content = append(content, interfaceBlock...)
		content = append(content, pubKeyPrefix...)
		content = append(content, s.PublicKey...)
		content = append(content, allowedIPPrefix...)
		content = append(content, endpoint...)
		content = append(content, endpointSuffix...)
		content = append(content, keepaliveBlock...)

		dj.Files = append(dj.Files, FileWrite{
			Name:    fname,
			Content: content,
		})
	}
}

func (g *Generator) writeJobsParallel(dirJobs []*DirJob, totalFiles int) error {
	if totalFiles == 0 {
		return nil
	}

	g.consoleManager.StartStatus("Preparing file system...")
	for _, dj := range dirJobs {
		if err := os.MkdirAll(dj.Path, 0755); err != nil {
			return err
		}
	}
	g.consoleManager.StopStatus()
	g.consoleManager.Success("File system prepared")

	g.consoleManager.StartProgress(totalFiles, "Writing configs")

	var completed int32
	var firstErr error
	var errMutex sync.Mutex

	uiDone := make(chan struct{})
	uiExited := make(chan struct{})

	go func() {
		defer close(uiExited)
		ticker := time.NewTicker(50 * time.Millisecond)
		defer ticker.Stop()
		for {
			select {
			case <-ticker.C:
				c := atomic.LoadInt32(&completed)
				g.consoleManager.UpdateProgress(int(c), totalFiles, "Writing configs")
			case <-uiDone:
				c := atomic.LoadInt32(&completed)
				g.consoleManager.UpdateProgress(int(c), totalFiles, "Writing configs")
				return
			}
		}
	}()

	workerCount := runtime.NumCPU() * 4
	if workerCount > 64 {
		workerCount = 64
	}
	if workerCount < 4 {
		workerCount = 4
	}
	if workerCount > len(dirJobs) {
		workerCount = len(dirJobs)
	}

	jobChan := make(chan *DirJob, len(dirJobs))
	for _, dj := range dirJobs {
		jobChan <- dj
	}
	close(jobChan)

	var wg sync.WaitGroup
	for i := 0; i < workerCount; i++ {
		wg.Add(1)
		go func() {
			defer wg.Done()
			for dj := range jobChan {
				for _, file := range dj.Files {
					fullPath := filepath.Join(dj.Path, file.Name)
					if err := os.WriteFile(fullPath, file.Content, 0644); err != nil {
						errMutex.Lock()
						if firstErr == nil {
							firstErr = err
						}
						errMutex.Unlock()
					}
					atomic.AddInt32(&completed, 1)
				}
			}
		}()
	}

	wg.Wait()
	close(uiDone)
	<-uiExited
	g.consoleManager.StopProgress()

	return firstErr
}

File: nordgen/src/nord_config_generator/generator.py
import asyncio
import os
from concurrent.futures import ThreadPoolExecutor
from dataclasses import dataclass
from datetime import datetime

from rich.progress import Progress, TaskID

from .client import NordClient
from .models import GenerationStats, Server, UserPreferences
from .server_parser import parse_servers
from .ui import ConsoleManager

_PATH_SANITIZE_TABLE: dict[int, int | None] = {ord(c): None for c in '<>:"/\\|?*\0'}
_PATH_SANITIZE_TABLE[ord(" ")] = ord("_")

_FILENAME_MAX_LENGTH = 15


@dataclass(slots=True, frozen=True)
class _ConfigWriteJob:
    absolute_path: str
    content: str


class Generator:
    def __init__(self, client: NordClient, ui: ConsoleManager) -> None:
        self.client = client
        self.ui = ui
        self.stats = GenerationStats()
        self.output_directory: str = ""

    async def process(
        self, private_key: str, preferences: UserPreferences
    ) -> str | None:
        self.output_directory = f"nordvpn_configs_{datetime.now().strftime('%Y%m%d_%H%M%S')}"

        with self.ui.status("Fetching data..."):
            (latitude, longitude), raw_servers = await asyncio.gather(
                self.client.get_geo(),
                self.client.get_servers(),
            )

        if not raw_servers:
            self.ui.fail("Failed to fetch server data")
            return None
        self.ui.success("Data fetched")

        with self.ui.status("Processing dataset..."):
            required_groups = set(preferences.groups) if preferences.groups else None

            all_parsed = parse_servers(
                raw_servers,
                latitude,
                longitude,
                required_groups=required_groups,
                exclude_dedicated=preferences.exclude_dedicated,
            )

            unique_servers = list({s.name: s for s in all_parsed}.values())
            
            if not unique_servers:
                self.ui.fail("No servers found matching the specified filters")
                return None
                
            unique_servers.sort(key=lambda s: (s.load, s.distance))

            self.stats.total = len(unique_servers)

            best_map: dict[tuple[str, str, str], Server] = {}
            for s in unique_servers:
                key = (s.combo, s.country, s.city)
                if key not in best_map:
                    best_map[key] = s
            self.stats.best = len(best_map)

            standard_jobs = self._build_jobs(
                unique_servers, "configs", private_key, preferences
            )
            best_jobs = self._build_jobs(
                list(best_map.values()), "best_configs", private_key, preferences
            )

        try:
            all_jobs = standard_jobs + best_jobs
            await asyncio.to_thread(
                self._materialize_directories, all_jobs
            )

            with self.ui.progress() as progress:
                task = progress.add_task("Writing all configs", total=len(all_jobs))
                await asyncio.to_thread(
                    self._write_jobs_parallel, all_jobs, progress, task
                )
        except OSError as err:
            self.ui.fail(f"Filesystem error: {err}")
            return None

        return self.output_directory

    def _build_jobs(
        self,
        servers: list[Server],
        subdirectory: str,
        private_key: str,
        preferences: UserPreferences,
    ) -> list[_ConfigWriteJob]:
        jobs: list[_ConfigWriteJob] = []
        counts: dict[str, int] = {}
        base_dir = self.output_directory

        interface_block = f"[Interface]\nPrivateKey = {private_key}\nAddress = 10.5.0.2/16\nDNS = {preferences.dns}\n\n[Peer]\n"
        keepalive_block = f"\nPersistentKeepalive = {preferences.keepalive}"

        for server in servers:
            country_seg = server.country.lower().translate(_PATH_SANITIZE_TABLE)
            city_seg = server.city.lower().translate(_PATH_SANITIZE_TABLE)
            fname_root = server.name.lower().translate(_PATH_SANITIZE_TABLE)[:_FILENAME_MAX_LENGTH]
            
            if not fname_root:
                fname_root = "unknown"

            directory = os.path.join(
                base_dir, subdirectory, server.combo, country_seg, city_seg
            )
            candidate = os.path.join(directory, f"{fname_root}.conf")

            count = counts.get(candidate, 0)
            counts[candidate] = count + 1
            final_path = candidate if count == 0 else os.path.join(directory, f"{fname_root}_{count}.conf")

            endpoint = server.station if preferences.use_ip else server.hostname
            content = f"{interface_block}PublicKey = {server.public_key}\nAllowedIPs = 0.0.0.0/0, ::/0\nEndpoint = {endpoint}:51820{keepalive_block}"
            jobs.append(_ConfigWriteJob(absolute_path=final_path, content=content))
        return jobs

    @staticmethod
    def _materialize_directories(jobs: list[_ConfigWriteJob]) -> None:
        unique_dirs = {os.path.dirname(job.absolute_path) for job in jobs}
        for d in unique_dirs:
            os.makedirs(d, exist_ok=True)

    @staticmethod
    def _write_jobs_chunk(jobs: list[_ConfigWriteJob], progress: Progress, task_id: TaskID) -> None:
        for job in jobs:
            with open(job.absolute_path, "w", encoding="utf-8") as f:
                f.write(job.content)
        progress.advance(task_id, len(jobs))

    @classmethod
    def _write_jobs_parallel(
        cls, jobs: list[_ConfigWriteJob], progress: Progress, task_id: TaskID
    ) -> None:
        chunk_size = 50
        cpu_count = os.cpu_count() or 1
        max_workers = min(64, max(4, cpu_count * 4))
        
        chunks = [jobs[i : i + chunk_size] for i in range(0, len(jobs), chunk_size)]
        
        with ThreadPoolExecutor(max_workers=max_workers) as executor:
            futures = [
                executor.submit(cls._write_jobs_chunk, chunk, progress, task_id)
                for chunk in chunks
            ]
            for future in futures:
                future.result()